src/address.rs

```diff
@@ -1,91 +1,338 @@
+#[derive(Clone, Eq, PartialEq, Debug)]
+pub struct Disconnected;
+
+/// General trait for any kind of address to an actor that is `Send`, be it strong or weak. This
+/// trait contains all functions of the address.
+#[doc(spotlight)]
+pub trait AddressExt<A: Actor> {
     /// Sends a [`Message`](trait.Message.html) that will be handled synchronously to the actor,
     /// and does not wait for a response. If this returns `Err(Disconnected)`, then the actor is stopped
     /// and not accepting messages. If this returns `Ok(())`, the will be delivered, but may
     /// not be handled in the event that the actor stops itself (by calling [`Context::stop`](struct.Context.html#method.stop))
     /// before it was handled.
-    pub fn do_send<M>(&self, message: M) -> Result<(), Disconnected>
+    fn do_send<M>(&self, message: M) -> Result<(), Disconnected>
+    where
+        M: Message,
+        A: Handler<M> + Send;
+
+    /// Sends a [`Message`](trait.Message.html) that will be handled asynchronously to the actor,
+    /// and does not wait for a response. If this returns `Err(Disconnected)`, then the actor is stopped
+    /// and not accepting messages. If this returns `Ok(())`, the will be delivered, but may
+    /// not be handled in the event that the actor stops itself (by calling [`Context::stop`](struct.Context.html#method.stop))
+    /// before it was handled.
+    fn do_send_async<M>(&self, message: M) -> Result<(), Disconnected>
+    where
+        M: Message,
+        A: AsyncHandler<M> + Send;
+
+    /// Sends a [`Message`](trait.Message.html) that will be handled asynchronously to the actor,
+    /// and waits for a response. If this returns `Err(Disconnected)`, then the actor is stopped
+    /// and not accepting messages.
+    fn send<M>(&self, message: M) -> MessageResponseFuture<M>
+    where
+        M: Message,
+        A: Handler<M> + Send,
+        M::Result: Send;
+
+    /// Sends a [`Message`](trait.Message.html) that will be handled asynchronously to the actor,
+    /// and waits for a response. If this returns `Err(Disconnected)`, then the actor is stopped
+    /// and not accepting messages.
+    fn send_async<M>(&self, message: M) -> MessageResponseFuture<M>
+    where
+        M: Message,
+        A: AsyncHandler<M> + Send,
+        for<'a> A::Responder<'a>: Future<Output = M::Result> + Send;
+
+    /// Attaches a stream to this actor such that all messages produced by it are forwarded to the
+    /// actor and handled synchronously. This could, for instance, be used to forward messages from a
+    /// socket to the actor (after the messages have been appropriately `map`ped). This is a
+    /// convenience method over a manual async stream forwarding loop.
+    #[doc(cfg(feature = "with-tokio-0_2"))]
+    #[doc(cfg(feature = "with-async_std-1"))]
+    #[cfg(any(doc, feature = "with-tokio-0_2", feature = "with-async_std-1"))]
+    fn attach_stream<S, M>(self, mut stream: S)
```

Comment:
You could possibly impl `Sink` for `T: AddrExt` which would allow `.forward` notation. Also then don't have to interact with tokio/async_std directly

Reply:
If that's not possible, what about making it an async fn that has to be executed by the caller? Removes the need for duplication and gives user more control :)

Reply:
A sink implementation might cause conflicting impl errors to arise (two impls have to exist - one for async handlers and one for sync handlers). However it would be possible to return an async fn but would also make it less ergonomic I feel? You'd have to call spawn on that manually. Otherwise I could implement both (like spawn vs create) but then there'd be four methods. Ultimately this is just a convenience wrapper so the flexibility can be achieved just with a few more lines of code using the public API

Reply:
Imo nicer to go with async fn. Is kinda weird to have library implicitly spawning onto global executor. Async fn is pretty standard anyway. `forward` returns a future, too

Reply:
I looked into this, and it isn't really possible without boxing (trait fns with async are boxed). It's just sugar over the direct thing so I'll leave that as a comment and then users can choose to do that on their own if they want to. It's all built with the public API, anyway.